KarhooUISDK/Screens/CheckoutScreen/Checkout/Workers/KarhooCheckoutBookingWorker.swift

```diff
@@ -327,33 +287,22 @@ final class KarhooCheckoutBookingWorker: CheckoutBookingWorker {
 
     // MARK: - Booking result handling
 
-    private func handleKarhooUserBookTripResult(_ result: Result<TripInfo>) {
-        guard let trip = result.getSuccessValue() else {
-            stateSubject.send(.idle)
-            reportBookingFailure(
-                message: result.getErrorValue()?.message ?? "",
-                correlationId: result.getCorrelationId()
-            )
-            if result.getErrorValue()?.type == .couldNotBookTripPaymentPreAuthFailed {
-                requestNewPaymentMethod(showRetryAlert: true)
-            } else {
-                stateSubject.send(.failure(result.getErrorValue() ?? ErrorModel.unknown()))
-            }
-
-            return
-        }
-
-        stateSubject.send(.success(trip))
-        reportBookingSuccess(tripId: trip.tripId, correlationId: result.getCorrelationId())
-    }
-
-    private func handleGuestAndTokenBookTripResult(_ result: Result<TripInfo>) {
+    private func handleBookTripResult(_ result: Result<TripInfo>) {
         if let trip = result.getSuccessValue() {
             reportBookingSuccess(tripId: trip.tripId, correlationId: result.getCorrelationId())
             stateSubject.send(.success(trip))
         } else if let error = result.getErrorValue() {
             reportBookingFailure(message: error.message, correlationId: result.getCorrelationId())
+            
+            // Show the error to the user
             stateSubject.send(.failure(result.getErrorValue() ?? ErrorModel.unknown()))
+            
+            // Adyen pre-auth usually fails before booking,
+            // while for Braintree we still get a nonce and the payment is refused after the booking call is made.
+            // For this scenario we need to reset the nonce and let the user restart the payment flow from scratch
+            if error.type == .couldNotBookTripPaymentPreAuthFailed {
```

Comment:
What would happen if we clear a nonce if any error occurs?

Reply:
The rest of the errors in the enum are not payment related judging by their names, so if we were to clear the nonce in their case we'd require the user to re-authorise the amount and we'd block the same amount as many times as there are attempts to book. Even if we unblock the amounts later, it's still bad UX in my opinion and could lead to confusion and mistrust.